In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
import os

import plotly.offline as pyo
pyo.init_notebook_mode(connected = True)

pd.options.plotting.backend = "plotly"

In [2]:
df = pd.read_excel("c:\\Users\\shaba\\PycharmProjects\\Data-Analysis\\Data\Online Retail.xlsx")

<>:1: SyntaxWarning:

invalid escape sequence '\O'

<>:1: SyntaxWarning:

invalid escape sequence '\O'

C:\Users\shaba\AppData\Local\Temp\ipykernel_11464\3107174252.py:1: SyntaxWarning:

invalid escape sequence '\O'



In [3]:
df.sample(5)

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
235769,557645,15036,ASSORTED COLOURS SILK FAN,7,2011-06-21 17:08:00,1.63,NaN,United Kingdom
428093,573413,23243,SET OF TEA COFFEE SUGAR TINS PANTRY,1,2011-10-30 16:12:00,4.95,15444.0,United Kingdom
325359,565437,22736,RIBBON REEL MAKING SNOWMEN,10,2011-09-04 13:48:00,1.65,17050.0,United Kingdom
374396,569374,23344,JUMBO BAG 50'S CHRISTMAS,1,2011-10-03 16:14:00,4.13,NaN,United Kingdom
456644,575703,22865,HAND WARMER OWL DESIGN,288,2011-11-10 16:59:00,1.85,13767.0,United Kingdom


In [4]:
df.shape

(541909, 8)

In [5]:
df.columns

Index(['InvoiceNo', 'StockCode', 'Description', 'Quantity', 'InvoiceDate',
       'UnitPrice', 'CustomerID', 'Country'],
      dtype='object')

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   InvoiceNo    541909 non-null  object        
 1   StockCode    541909 non-null  object        
 2   Description  540455 non-null  object        
 3   Quantity     541909 non-null  int64         
 4   InvoiceDate  541909 non-null  datetime64[ns]
 5   UnitPrice    541909 non-null  float64       
 6   CustomerID   406829 non-null  float64       
 7   Country      541909 non-null  object        
dtypes: datetime64[ns](1), float64(2), int64(1), object(4)
memory usage: 33.1+ MB


In [7]:
df.isnull().sum()

InvoiceNo           0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     135080
Country             0
dtype: int64

In [8]:
df.duplicated().sum()

np.int64(5268)

In [9]:
df.describe()

,Quantity,InvoiceDate,UnitPrice,CustomerID
count,541909.000000,541909,541909.000000,406829.000000
mean,9.552250,2011-07-04 13:34:57.156386048,4.611114,15287.690570
min,-80995.000000,2010-12-01 08:26:00,-11062.060000,12346.000000
25%,1.000000,2011-03-28 11:34:00,1.250000,13953.000000
50%,3.000000,2011-07-19 17:17:00,2.080000,15152.000000
75%,10.000000,2011-10-19 11:27:00,4.130000,16791.000000
max,80995.000000,2011-12-09 12:50:00,38970.000000,18287.000000
std,218.081158,NaN,96.759853,1713.600303


## Quality
- Missing values in Description and CustomerID
- CustomerID datatype is float and InvoiceDate datatye is datetime64[ns]
- Negative values in quantity
- inconsistency in InvoiceNo and StockCode i.e. contains alphabets at start in some InvoiceNo and at end In some StockCode 

## Tidiness

- Same column for date and time i.e. InvoiceDate
- 5268 duplicate rows

In [10]:
data = df.copy()

#### Handling Missing Values

In [11]:
print(f"missing_Description' = {(data.isnull().sum()['Description']/len(data)) *100:.2f} %")

missing_Description' = 0.27 %


In [12]:
print(f"missing_CustomerID = {(data.isnull().sum()['CustomerID']/len(data)) *100:.2f} %")

missing_CustomerID = 24.93 %


In [13]:
len(df['Description'].dropna())/ len(df)

0.9973168926886249

In [14]:
new_df = df['Description'].dropna()
df.shape, new_df.shape

((541909, 8), (540455,))

In [ ]:
df['Quantity'].plot()

ValueError: Mime type rendering requires nbformat>=4.2.0 but it is not installed